New rule: Link is descriptive (5effbb) (#555)
act-rules/act-rules.github.io: pages/glossary/programatically-determined-link-context.md

--

> title: Programmatically Determined Link Context

Reviewer: I think you're making an assumption here that HTML is valid. Let's take something like this:

```html
<p>
  <h1>ACT Rules Community Group</h1>
  <p><a href="...">learn more</a></p>
</p>
```

I would argue that this should fail the SC, but according to this definition, the heading text is part of the context, and so it passes. I don't know if that's a bad thing necessarily. This doesn't cause false positives, but it doesn't tell the whole story either. I'm OK ignoring this edge case, but if we do, I think there needs to be an assumption.

Author: I agree that this should fail the SC. I've updated the definition to look only at the nearest ancestor of the link element, so the heading is no longer part of the context.

Reviewer: There's a lot of repetition in this new text. You've also only solved the case of the element being the same, But similarly to the previous example, I could swap out the top level `p` for another element in your list, and you'd have the same problem again:

```html
<li>
  <h1>ACT Rules Community Group</h1>
  <p><a href="...">learn more</a></p>
</li>
```

Additionally, this no longer allow nested list items to be part of context, like you have in passed example 5. This would now fail the rule:

```html
<ul>
  <li> Ulysses <ul>
    <li><a href="...">HTML</a></li>
    <li><a href="...">EPUB</a></li>
    <li><a href="...">Plain text</a></li>
  </ul></li>
</ul>
```

Author: Regarding the first example, I now would say that this counts as context for this SC (i.e., the text from the heading is in the same listitem). It is my interpretation of the understanding document that context comes from the nearest ancestor (otherwise it should have been written "includes text that is in the same paragraph**s**, list**s**, or table cell**s**...")

This interpretation does have the side effect of now failing passed example 5 (nicely spotted). I've updated it to use the aria-describedby attribute to provide the required context.

I don't really know what to do about the text repetition. I wasn't very happy with it, but this was the most succint I managed to be. Any suggestions to improve it, @WilcoFiers?

Author: Thanks for the suggestion. In it you are listing elements and not relationships, so I've updated it to list relationships.

key: programmatically-determined-link-context
--

> In HTML, information that is programmatically determinable from a link includes [text nodes][] that are [included in the accessibility tree][] with a non-empty ("") [accessible name][] and are

Reviewer: I don't think this should be limited to text nodes. That way you're ignoring images and other things with accessible names through attributes.

> - in the same paragraph (HTML `p` element) as the link;

Reviewer: Need to clarify what "in the same" means. I suspect what you mean is that it has the same `p` element as an ancestor in the flattened tree, correct?

> - in the same list (HTML element with the semantic role of `listitem`) as the link;

Reviewer: This is confusing. Why not say in the same listitem, if that's the role you're after.

> - sibling nodes in the same table cell (HTML element with the semantic role of `cell` or `gridcell`) as the link;

Reviewer: I don't think sibling nodes are sufficient here. For example:

```html
<p>Go to the <span class="coolStypes">Wikipedia</span> <a href="http://wikipedia.org/">homepage</a></p>
```

In this case the text node "Wikipedia" is not a sibling node of the link. Or alternatively you could have something like this:

```html
<p>Go to the Wikipedia <u><a href="http://wikipedia.org/">homepage</u></a></p>
```

Since the link is wrapped in `u`, the link has no sibling nodes.

> - in a table header cell (HTML element with the semantic role of `columnheader` or `rowheader`) that is [associated][] with the table cell (HTML element with the semantic role of `cell` or `gridcell`) of which the link is a descendent;

Reviewer: Descendent in the flat tree I'm guessing. You may want to create a definition of "descendent" that specifies it's a descendent in the flat tree.

Author: Updated, but I don't think the definition is needed

> - nodes of elements referenced by an `aria-describedby` attribute of the link element.

Reviewer 2: I know it's how the WCAG def is written, but I'm not super happy about dropping headings from the context.

Reading technique ARIA7, I kinda guess that the idea is to explicitly include them through a `aria-describedby` or `aria-labelled-by` attribute when needed.

Maybe it's worth illustrating this by adding examples to the definition (like in #994).

Author: Headings should still be included in the programmatically determined context definition. It's just that the programmatically determined link context definition is stricter.

--
title: Programmatically Determined Link Context
key: programmatically-determined-link-context
--

> The programmatically determined link context is a list of elements. All elements in the list must be [included in the accessibility tree][], and have one of the following relationships to the link:

Reviewer 2: I have to point out that "the closest ancestor that is a `p`" is not a relationship _stricto sensu_… (the relationship would be "**is** the closest `p` ancestor"…)

Plus, the definition is currently ambiguous because it doesn't specify the size of the list. So, there is no unicity of PDL context and, typically, the empty list is always a PDL context for any link.

---
The PDL context is the list of all elements which are included in the accessibility tree and are either:
- an ancestor of the link with a role of listitem; or
- the closest ancestor of the link that is a `p`; or
- …

should be good.

---
Replace "serve as row/column header for the cell" by "is assigned to the cell containing the link" + link to https://html.spec.whatwg.org/multipage/tables.html#algorithm-for-assigning-header-cells

> - sibling [phrasing content][] nodes in the same [paragraph][] as the link;

Reviewer: Phrasing content doesn't take text alternatives into consideration. Nor does it consider pseudo element content. I'm not sure what the best way to go here is. Perhaps you can look at the accessible name of these elements, or if they don't have one the name from contents value ([Step 2F.i in accName 1.1](https://www.w3.org/TR/accname-1.1/#step2F.i)).

Author: The definition requires that nodes contributing to the programmatically determined context are text nodes included in the accessibility tree. I'll add to those requirements that they have a non-empty accessible name. With this, it is safe to remove the phrasing content requirement and leave it as "sibling nodes in the same paragraph as the link"

> - sibling [phrasing content][] nodes in the same [paragraph][] as the link;

Reviewer: I don't think paragraph works here. I wager all screen readers look at `p` elements for paragraphs, not this HTML spec definition of it.

- sibling [phrasing content][] nodes in the same list (HTML element with the semantic role of `listitem`) as the link;

> - sibling [phrasing content][] nodes in the same table cell (HTML element with the semantic role of `cell` or `gridcell`) as the link;

Reviewer: What if a link is contained inside a list inside a paragraph inside a table cell? Do you get to use all of them as context? I would argue it only counts towards the closest one.

Author: I don't agree. I think all should be used as context. In the example below don't you think that the table header and the text in the same cell are providing context even if the element is inside a list in the cell?

```html
<table>
  <tr>
    <th colspan="3">Ulysses</th>
  </tr>
  <tr>
    <td>
      Download the book in your preferred format:
      <ul>
        <li>
          <a href="https://www.gutenberg.org/files/4300/4300-h/4300-h.htm">
            HTML
          </a>
        </li>
        <li>
          <a
            href="https://www.gutenberg.org/ebooks/4300.epub.images?session_id=04cd710372888de8d8d322215cdfe8ce5b0f8d73"
          >
            EPUB
          </a>
        </li>
        <li>
          <a href="https://www.gutenberg.org/files/4300/4300-0.txt">
            Plain text
          </a>
        </li>
      </ul>
    </td>
  </tr>
</table>
```

--
title: Programmatically Determined Link Context
key: programmatically-determined-link-context
--

> The programmatically determined link context is a list of elements. All elements in the list must be [included in the accessibility tree][], and have one of the following relationships to the link:

Reviewer 2: There is still an issue on how many elements matching the condition are in the Programmatically Determined Link Context, making the definition ambiguous.

```html
<span id="label>Foo 1</span>
<p>Foo 2
<ul>
<li>Foo 3
<table>
<tr><th>Foo 4</th></tr>
<tr><td>Foo 5 <a href="#" aria-described-by="label">Bar</a></td></tr>
</table>
</ul>
</p>
```
What is the programmatically determined link context of this link?
With the current definition, any subset of {Foo 1, Foo 2, Foo 3, Foo 4, Foo 5} (including the empty set!) can be.

---
Also,  "list" kinda suggest ordering of the elements in the context, which is probably not what you want. "set" seems better.

---
```suggestion
The programmatically determined link context is the set of all elements that are [included in the accessibility tree][], and have one of the following relationships to the link:
```

--
title: Programmatically Determined Link Context
key: programmatically-determined-link-context
--

In HTML, information that is programmatically determinable from a link includes HTML elements that are [visible][] and [included in the accessibility tree][] with a non-empty ("") [accessible name][] and

> - have as ancestor in the [flat tree][] a `p` element that is [ancestor][] of the link;

Author: Solved by your suggestion 👍
